tests/test_oidc.py

```diff
@@ -0,0 +1,138 @@
+    'token_type': 'bearer'}
+
+
+@pytest.fixture
+def claims(utcnow):
+    config = {
+        "issuer": "http://dex.example.com:5556",
+        "authorization_endpoint": "http://dex.example.com:5556/auth",
+        "token_endpoint": "http://dex.example.com:5556/token",
+        "jwks_uri": "http://dex.example.com:5556/keys",
+        "response_types_supported": [
+            "code"
+        ],
+        "grant_types_supported": [
+            "authorization_code",
+            "client_credentials"
+        ],
+        "subject_types_supported": [
+            "public"
+        ],
+        "id_token_signing_alg_values_supported": [
+            "RS256"
+        ],
+        "token_endpoint_auth_methods_supported": [
+            "client_secret_basic"
+        ],
+        "redirect_uri": "test",
+        "client_id":
+        "8prYoZym0zkt6ZzTIbXaFBQvtrWi2A8xmdC9WEz3i14=@localhost",
+            "client_secret": "testsecret",
+            "keys": [
+                {
+                    "kid": "4fb1565eade51b9c3a52be4424b3ca91c38f5365",
+                    "kty": "RSA",
+                    "alg": "RS256",
+                    "use": "sig",
+                    "e": "AQAB",
+                    "n": (
+                        "zzJAqac1b3GVEehBZUp48gjJpXkD-vGsFyeTmWJGl00NVmgimW-h7"
+                        "UUrVtJiwnaXS98D3ZPFnQRUrn1bmLeADkHs2sbWyfCNTYVK2sBY0Y"
+                        "K6AamIwCLEET8aWB2b0P6L7xpM8lP8y3ss42ICgWV73txv-KwI8Fw"
+                        "aDScIzSzSSdWWgqxh54PCMGAbWJCg6kjyufBocBVaZLUL_4HKIixI"
+                        "SgyThUpFYND8iaQThI6fkW5o3Qc95AEcm5XKvlEXYv-BWWj-xBXka"
+                        "FDy9YOjUKAXcYmWPhbY5LGM1GYk79PigNggxgRDsrxWDNQUzH25o8"
+                        "wxRfOX1cZ-bdVzGYHDYeUbpw=="
+                    )
+                }
+        ]
+    }
+    now = datetime.datetime(2016, 9, 14, 12, 40)
+    with utcnow("jose.jwt.datetime", now):
+        claims = verify_id_token(test_token_response['id_token'], config)
+
+    return claims
+
+
+@pytest.yield_fixture
+def mock_auth(claims):
+
+    with mock.patch("app.blueprints.base.views.oidc") as oidc:
+        oidc.authenticate.return_value = claims
+        yield
+
+
+@pytest.yield_fixture
+def ds_spy():
+    with mock.patch("app.extensions.user_datastore") as ds:
+        ds.create_user = mock.Mock()
+        yield ds.create_user
+
+
+class WhenClientReceivesAnIDToken(object):
+
+    def it_contains_an_auth_time_claim(self, claims):
+        assert "iat" in claims
+
+
+class WhenUserIsAuthenticated(object):
+
+    def it_stores_issuer_and_subject_id(self, mock_auth, client, db_session):
+
+        url = url_for("base.oidc_callback")
+
+        response = client.get(url)
+        assert response.status_code == 302
+
+        users = User.query.filter_by(
+            email='ken.tsang@digital.cabinet-office.gov.uk',
+            issuer_id='http://dex.example.com:5556',
+            subject_id='6c7a9f45-7f57-4d91-be0e-286673a28c6f'
```

Comment:
These values are a bit "magic number"-ish - maybe it would be more self-explanatory, if we put these in a `expected_claims` variable and then had something like:
```python
users = User.query.filter_by(
    email=expected_claims['email'],
    issuer_id=expected_claims['iss'],
    subject_id=expected_claims['sub']).all()
```